src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/CustomExportAssemblyLoader.cs

```diff
@@ -2,98 +2,95 @@
+    /// Loads assemblies from either the host or from our extensions.
+    /// If an assembly exists in both the host and an extension, we will use the host assembly for the MEF catalog.
     /// </summary>
-    private readonly string _baseDirectory;
+    public Assembly LoadAssembly(AssemblyName assemblyName)
+    {
+        // VS-MEF generally tries to populate AssemblyName.CodeBase with the path to the assembly being loaded.
+        // We need to read this in order to figure out which ALC we should load the assembly into.
+#pragma warning disable SYSLIB0044 // Type or member is obsolete
+        var codeBasePath = assemblyName.CodeBase;
+#pragma warning restore SYSLIB0044 // Type or member is obsolete
+        return LoadAssembly(assemblyName, codeBasePath);
+    }
 
-    public CustomExportAssemblyLoader(string baseDirectory)
+    public Assembly LoadAssembly(string assemblyFullName, string? codeBasePath)
     {
-        _baseDirectory = baseDirectory;
+        var assemblyName = new AssemblyName(assemblyFullName);
+        return LoadAssembly(assemblyName, codeBasePath);
     }
 
-    public Assembly LoadAssembly(AssemblyName assemblyName)
+    private Assembly LoadAssembly(AssemblyName assemblyName, string? codeBasePath)
     {
-        Assembly? value;
-        lock (_loadedAssemblies)
+        _logger.LogTrace($"Loading assembly {assemblyName}");
+        // First attempt to load the assembly from the default context.
+        Exception loadException;
+        try
         {
-            _loadedAssemblies.TryGetValue(assemblyName, out value);
+            return AssemblyLoadContext.Default.LoadFromAssemblyName(assemblyName);
         }
-
-        if (value == null)
+        catch (FileNotFoundException ex)
         {
-            // Attempt to load the assembly normally, but fall back to Assembly.LoadFrom in the base
-            // directory if the assembly load fails
-            try
-            {
-                value = Assembly.Load(assemblyName);
-            }
-            catch (FileNotFoundException) when (assemblyName.Name is not null)
-            {
-                var filePath = Path.Combine(_baseDirectory, assemblyName.Name)
-                    + (assemblyName.Name.EndsWith(".dll", StringComparison.OrdinalIgnoreCase)
-                        ? ""
-                        : ".dll");
-
-                value = Assembly.LoadFrom(filePath);
-
-                if (value is null)
-                {
-                    throw;
-                }
-            }
+            loadException = ex;
+            // continue checking the extension contexts.
+        }
 
-            lock (_loadedAssemblies)
-            {
-                _loadedAssemblies[assemblyName] = value;
-                return value;
-            }
+        if (codeBasePath is not null)
+        {
+            return LoadAssemblyFromCodeBase(assemblyName, codeBasePath);
         }
```

Comment:
Random thought: if we have a codebase path and we know the path isn't our application direectory, should have just done this first? Rather than that earlier load that we could have predicted would fail?